File: Algos_on_Strings/bwtinverse.py
# python3
import sys

def InverseBWT(bwt):
    # Get first and last columns    
    last = list(bwt)
    first = sorted(list(last))
    # Labels symbols with a ctr to when the number appeared
    first, last = LabedSymbols(first), LabedSymbols(last)
    # Denote first to last
    D = {f:last[i] for i,f in enumerate(first)}
    key = D[('$', 1)]
    text = [key[0]]
    # find origonal text
    while key != ('$', 1):
        key = D[key]
        text.append(key[0])
    return ''.join(text[:-1][::-1]) + '$'

def LabedSymbols(values):
    symbols = [ ]
    cnt = { }
    for v in values:
        if not v in cnt.keys():
            cnt[v] = 1
        c = cnt[v]
        cnt[v] = c + 1
        symbols.append((v, c))
    return symbols

if __name__ == '__main__':
    #bwt = sys.stdin.readline().strip()
    bwt = 'annb$aa'
    #bwt = 'AC$A'
    #bwt = 'AGGGAA$'
    print(InverseBWT(bwt))

File: Algos_on_Strings/bwtinverse.cs
using System;
using System.Collections.Generic;
using System.Linq;

    class Program
    {
        static string InverseBWT(string bwt)
        {
            var lastVals = bwt.ToList();
            var firstVals = lastVals.OrderBy(v => v).ToList();
            var first = LabeledSymbols(firstVals);
            var last = LabeledSymbols(lastVals);
            var D = first.Select((f, i) => new KeyValuePair<Tuple<char, int>, Tuple<char, int>>(f, last[i]))
                        .ToDictionary(s => s.Key, s => s.Value);
            var startKey = new Tuple<char, int>('$', 1);
            var key = D[startKey];
            var text = new List<char> { key.Item1 };
            while (key.Item1 != startKey.Item1 || key.Item2 != startKey.Item2)
            {
                key = D[key];
                text.Add(key.Item1);
            }
            return new string(text.Take(text.Count - 1).Reverse().ToArray()) + "$";
        }
        
        static List<Tuple<char, int>> LabeledSymbols(IEnumerable<char> values)
        {
            var symbols = new List<Tuple<char, int>>();
            var cnt = new Dictionary<char, int>();
            foreach(var v in values)
            {
                if (!cnt.ContainsKey(v))
                    cnt[v] = 1;
                var c = cnt[v];
                cnt[v]++;
                symbols.Add(new Tuple<char, int>(v, c));
            }
            return symbols;
        }

        
        static void Main(string[] args)
        {
            var inp = Console.ReadLine();
            //var inp = "annb$aa";
            //inp = "AGGGAA$";
            Console.Write(InverseBWT(inp));
        }
    }
